presto-main/src/main/java/com/facebook/presto/operator/scalar/KleinbergBurstDetectionFunction.java

```diff
@@ -0,0 +1,225 @@
+    private static Block kleinberg(Block timeSeries, double s, double gamma)
+    {
+        // The number of events [N].
+        int numEntries = timeSeries.getPositionCount();
+
+        // Calculate time intervals between successive events.
+        double[] intervals = new double[numEntries - 1];
+        // The minimum interval.
+        double delta = Double.MAX_VALUE;
+        for (int i = 0; i < numEntries - 1; i++) {
+            double timeCurrent = DOUBLE.getDouble(timeSeries, i);
+            double timeNext = DOUBLE.getDouble(timeSeries, i + 1);
+            // 'timeSeries' should be sorted in ascending order.
+            checkCondition(timeNext > timeCurrent, INVALID_FUNCTION_ARGUMENT, "time series must be sorted in ascending order");
+            intervals[i] = timeNext - timeCurrent;
+            delta = Math.min(delta, intervals[i]);
+        }
+
+        // The time length of the whole timeSeries [T].
+        double timeLength = DOUBLE.getDouble(timeSeries, numEntries - 1) - DOUBLE.getDouble(timeSeries, 0);
+        checkCondition(timeLength > 0, INVALID_FUNCTION_ARGUMENT, "time length of the whole series must be greater than 0");
+
+        // The upper limit of burst levels [K].
+        int burstLevelUpperLimit = (int) (Math.ceil(1 + Math.log(timeLength / delta) / Math.log(s)));
+        checkArgument(burstLevelUpperLimit > 0);
+
+        double[] alpha = new double[burstLevelUpperLimit];  // Event rate at each burst level.
+        double[] alphaLn = new double[burstLevelUpperLimit];
+
+        alpha[0] = numEntries / timeLength;  // Average event rate.
+        alphaLn[0] = Math.log(alpha[0]);
+
+        for (int i = 1; i < burstLevelUpperLimit; i++) {
+            alpha[i] = s * alpha[i - 1];
+            alphaLn[i] = Math.log(alpha[i]);
+        }
+
+        // Initialize.
+        Node[] q = new Node[burstLevelUpperLimit];
+        double[] costs = new double[burstLevelUpperLimit];
+        Arrays.fill(costs, Double.MAX_VALUE);
+        costs[0] = 0;
+
+        // Start optimization.
+        double gammaLogN = gamma * Math.log((double) numEntries);
+        for (double interval : intervals) {
+            Node[] qNew = new Node[burstLevelUpperLimit];
+            double[] costsNew = new double[burstLevelUpperLimit];
+            for (int i = 0; i < burstLevelUpperLimit; i++) {
+                double[] c = new double[burstLevelUpperLimit];
+                for (int j = 0; j < burstLevelUpperLimit; j++) {
+                    c[j] = costs[j] + tau(gammaLogN, j, i);
+                }
+                int jMin = findIndexOfSmallest(c);
+                // Store the cost for setting the burst level i.
+                costsNew[i] = -logf(alpha, alphaLn, i, interval) + c[jMin];
+                qNew[i] = new Node(i, q[jMin]);
+            }
+            q = qNew;
+            costs = costsNew;
+        }
+
+        // Build output structure.
+        List<Integer> bursts = new ArrayList<>(nCopies(numEntries, 0));
```

Comment:
Let's use an array here, as it will avoid boxing. The usual reasons to prefer a list don't apply here, as it has a fixed size, doesn't escape the function, and is used exactly like an array.

This can be
```java
int[] bursts = new int[numEntries];
```

The default is to initialize with zero (the default value for the type).